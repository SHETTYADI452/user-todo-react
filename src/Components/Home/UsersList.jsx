//Importing modules
import React from "react";

//exporting allusers and onclick to  index.jsx so that we can map the data of allUsers and perform a click event.
export const UsersList = ({ allUsers, onUserClick }) => {
  // will return a list of users
  return (
    <ul className="users-list-container">
      <ul>
        {allUsers?.map((user, index) => (
          <p
            key={`user-${index}`}
            className={`user-${index}`}
            id={`user-${index}`}
            onClick={() => onUserClick(user)}
          >
            {user.id}. {user.username}
          </p>
        ))}
      </ul>
    </ul>
  );
};
