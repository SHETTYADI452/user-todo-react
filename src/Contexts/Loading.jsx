// importing modules
import { createContext, useReducer } from "react";
import { Loading, initialState } from "../reducers/loading";

// creating loading context
export const LoadingContext = createContext({
  state: initialState,
  dispatch: () => null,
});

// creating loading context provider
export const LoadingContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(Loading, initialState);
  const value = { state, dispatch };

  return (
    <LoadingContext.Provider value={value}>{children}</LoadingContext.Provider>
  );
};
