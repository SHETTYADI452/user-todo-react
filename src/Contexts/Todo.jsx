// importing modules
import { createContext, useReducer } from "react";
import { Todo, initialState } from "../reducers/todo";

// creating todo context
export const TodoContext = createContext({
  state: initialState,
  dispatch: () => null,
});

//creating todo context provider
export const TodoContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(Todo, initialState);
  const value = { state, dispatch };

  return <TodoContext.Provider value={value}>{children}</TodoContext.Provider>;
};
