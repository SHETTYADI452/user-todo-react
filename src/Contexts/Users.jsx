// importing modules
import { createContext, useReducer } from "react";
import { Users, initialState } from "../reducers/users";

// creating context provider
export const UsersContext = createContext({
  state: initialState,
  dispatch: () => null,
});

// creating  Users context provider
export const UsersContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(Users, initialState);
  const value = { state, dispatch };

  return (
    <UsersContext.Provider value={value}>{children}</UsersContext.Provider>
  );
};
