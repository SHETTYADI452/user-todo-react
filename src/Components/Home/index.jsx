// importing modules
import React, { useEffect, useContext } from "react";
import { fetchUsers, setSelectedUser } from "../../actions/users";
import { UsersList } from "./UsersList";
import { UsersContext } from "../../Contexts/Users";
import { LoadingContext } from "../../Contexts/Loading";
import { useHistory } from "react-router-dom";
import ClipLoader from "react-spinners/ClipLoader";
import "../../styles/home.css";

//export default class
export default (props) => {
  const title = "USERS LIST";
  const history = useHistory();
  //Initializing users context
  const { state: userState, dispatch: userDispatch } = useContext(UsersContext);
  // Initializing loading context
  const { state: loadingState, dispatch: loadingDispatch } =
    useContext(LoadingContext);

  // with the help of useEffect Hook the as soons as the app renders the userdetails will be loaded with effect
  useEffect(() => {
    fetchUsers()(userDispatch, loadingDispatch);
  }, []);

  // This function will help to fetch user details with the help of diapatch and Navigate users to userdeatails page
  const handleOnUserClick = (user) => {
    // this will push to new page
    history.push(`/userdetail-todo/${user.id}`);
    // this function will invoke in action/user folder and set the userData
    setSelectedUser(user)(userDispatch, loadingDispatch);
  };

  return (
    <div>
      <center>
        <div id="page-title">
          <h1>{title}</h1>
        </div>

        <div className="userlist">
          {loadingState.isLoading ? (
            <div>
              <h3 className="loader">Fetching Users...</h3>
              <ClipLoader loading={loadingState.isLoading} size={150} />
            </div>
          ) : (
            <UsersList
              allUsers={userState.users}
              onUserClick={handleOnUserClick}
            />
          )}
        </div>
      </center>
    </div>
  );
};
