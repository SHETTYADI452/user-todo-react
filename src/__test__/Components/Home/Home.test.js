import { mount, shallow } from "enzyme";
import Home from "../../../Components/Home";
import { UsersContext } from "../../../Contexts/Users";
import { LoadingContext } from "../../../Contexts/Loading";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const allUsers = [
  {
    id: 1,
    username: "user 1",
  },
  {
    id: 2,
    username: "user 2",
  },
  {
    id: 3,
    username: "user 3",
  },
];

describe("Testing Home component", () => {
  let wrapper;
  let secondaryWrapper;

  const title = "USERS LIST";

  beforeEach(() => {
    wrapper = mount(
      <UsersContext.Provider
        value={{
          state: {
            users: allUsers,
            user: {},
          },
          dispatch: jest.fn(),
        }}
      >
        <Home />
      </UsersContext.Provider>
    );
  });
  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should match the snapshot - shallow", () => {
    wrapper = shallow(<Home title={title} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should match the snapshot having loader while passing loading as true in the Loading Context", () => {
    secondaryWrapper = mount(
      <LoadingContext.Provider
        value={{
          state: { isLoading: true },
          dispatch: jest.fn(),
        }}
      >
        <UsersContext.Provider
          value={{
            state: {
              users: allUsers,
              user: {},
            },
            dispatch: jest.fn(),
          }}
        >
          <Home />
        </UsersContext.Provider>
      </LoadingContext.Provider>
    );

    expect(secondaryWrapper).toMatchSnapshot();
  });

  it("should have a page title with text Home", () => {
    const pageTitleElement = wrapper.find("#page-title");
    expect(pageTitleElement.text()).toBe(title);
  });

  it("should find the user 1 element from users list and clicking on it", () => {
    expect(wrapper.find(`#user-0`).length).toBe(1);
    wrapper.find(`#user-0`).simulate("click");
  });
});
