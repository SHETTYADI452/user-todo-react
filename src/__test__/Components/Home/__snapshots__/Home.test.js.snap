// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home component should match the snapshot - shallow 1`] = `ShallowWrapper {}`;

exports[`Testing Home component should match the snapshot 1`] = `ReactWrapper {}`;

exports[`Testing Home component should match the snapshot having loader while passing loading as true in the Loading Context 1`] = `ReactWrapper {}`;
